narwhals/_ibis/namespace.py

```diff
@@ -0,0 +1,268 @@
+from __future__ import annotations
+
+import operator
+from functools import reduce
+from itertools import chain
+from typing import TYPE_CHECKING
+from typing import Any
+from typing import Iterable
+from typing import Sequence
+from typing import cast
+
+import ibis
+import ibis.expr.types as ir
+
+from narwhals._compliant import CompliantThen
+from narwhals._compliant import LazyNamespace
+from narwhals._compliant import LazyWhen
+from narwhals._expression_parsing import combine_alias_output_names
+from narwhals._expression_parsing import combine_evaluate_output_names
+from narwhals._ibis.dataframe import IbisLazyFrame
+from narwhals._ibis.expr import IbisExpr
+from narwhals._ibis.selectors import IbisSelectorNamespace
+from narwhals._ibis.utils import lit
+from narwhals._ibis.utils import narwhals_to_native_dtype
+from narwhals.utils import Implementation
+from narwhals.utils import requires
+
+if TYPE_CHECKING:
+    from narwhals.dtypes import DType
+    from narwhals.typing import ConcatMethod
+    from narwhals.utils import Version
+
+
+class IbisNamespace(LazyNamespace[IbisLazyFrame, IbisExpr, "ir.Table"]):
+    _implementation: Implementation = Implementation.IBIS
+
+    def __init__(self, *, backend_version: tuple[int, ...], version: Version) -> None:
+        self._backend_version = backend_version
+        self._version = version
+
+    @property
+    def selectors(self) -> IbisSelectorNamespace:
+        return IbisSelectorNamespace.from_namespace(self)
+
+    @property
+    def _expr(self) -> type[IbisExpr]:
+        return IbisExpr
+
+    @property
+    def _lazyframe(self) -> type[IbisLazyFrame]:
+        return IbisLazyFrame
+
+    def concat(
+        self, items: Iterable[IbisLazyFrame], *, how: ConcatMethod
+    ) -> IbisLazyFrame:
+        if how in {"horizontal", "diagonal"}:
+            msg = f"{how!r} concat not supported for Ibis. Please join instead."
+            raise NotImplementedError(msg)
```

Comment:
`'horizontal'` is already disallowed before we get here, if we can't support 'diagonal' let's just raise on that

Reply:
Updated to only raise on `how == "diagonal"`